### PR TITLE
 event: adding the capability of creating an alarm with a given scope

### PR DESCRIPTION
**Pull request by alyssawilk**
Precursor to https://github.com/envoyproxy/envoy/pull/7782
Adding scope tracking functionality to the basic alarm functions.

Risk Level: Medium (should be a no-op but is a large enough refactor)
Testing: new unit tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: include/envoy/event/dispatcher.h [hunk 1]
```diff
@@ -149,8 +149,9 @@ class Dispatcher {
   /**
    * Allocate a timer. @see Timer for docs on how to use the timer.
    * @param cb supplies the callback to invoke when the timer fires.
+   * @param cb supplies an optional scope for the duration of the alarm.
    */
-  virtual Event::TimerPtr createTimer(TimerCb cb) PURE;
+  virtual Event::TimerPtr createTimer(TimerCb cb, const ScopeTrackedObject* object = nullptr) PURE;
```
Comment (mattklein123) [inline include/envoy/event/dispatcher.h]: High level question: Do you think we should pass the scope as part of timer creation or rather when we enable the timer? My concern is I wonder if we might at some point have a different scope for different deadlines on the same timer? TBH this seems unlikely, but at the same time I worry about having to redo this interface if/when that happens? WDYT?

/wait-any

Comment (alyssawilk) [reply include/envoy/event/dispatcher.h]: Should that go for passing the dispatcher through as well, or should that stay in the constructor?

Comment (mattklein123) [reply include/envoy/event/dispatcher.h]: Sorry what do you mean?

Comment (alyssawilk) [reply include/envoy/event/dispatcher.h]: Currently in the constructor we're passing through the dispatcher (also needed for scope setting) and the scope.
Preference for moving both to enableTimer, or leaving dispatcher as a one-time thing?

Comment (mattklein123) [reply include/envoy/event/dispatcher.h]: IMO the dispatcher should not change so constructor is fine. WDYT?

Comment (alyssawilk) [reply include/envoy/event/dispatcher.h]: doesn't change today, and if we ever have alarms that switch threads I think the constructor is scoped enough we can do the work then.

**Comment by alyssawilk** [comment]
Ok, I see what's causing the failure.
Various threads can dispatcher->post() callbacks to be run in a worker thread, at the same time.
dispatcher->post() enables the post_timer_.  Because this can be called by two threads simultaneously, they both edit the scope simultaneously, and cause the tsan failure.

I can hack around this by only setting object_ if it's non-null but that's terrible.  Fundamentally, I think enabling the post timer should not be required to be thread safe, so we should acquire some lock and enable the timer if it's not already enabled in dispatcher()->post.  I assume there's some good reason setting the post timer isn't already under the post lock, but I'm not seeing it
@mattklein123

**Comment by mattklein123** [comment]
> I assume there's some good reason setting the post timer isn't already under the post lock, but I'm not seeing it

I think it's just because it never needed to before and libevent deals with this via its own lock in the event base. I can't think of any reason why it wouldn't be OK to acquire a lock there. The alternative would be to just use an atomic pointer to the scope and that I think would also yield the same behavior?

**Comment by alyssawilk** [comment]
fair point.   I think I'd rather move it under the lock but I'm mildly concerned about potential deadlock so I'll make it an atomic for now and maybe follow up with the smaller (but higher risk) change.

File: source/common/event/timer_impl.cc [hunk 2]
```diff
@@ -17,16 +17,28 @@ void TimerUtils::millisecondsToTimeval(const std::chrono::milliseconds& d, timev
...
   evtimer_assign(
       &raw_event_, libevent.get(),
-      [](evutil_socket_t, short, void* arg) -> void { static_cast<TimerImpl*>(arg)->cb_(); }, this);
+      [](evutil_socket_t, short, void* arg) -> void {
+        TimerImpl* timer = static_cast<TimerImpl*>(arg);
+        if (timer->object_ == nullptr) {
+          timer->cb_();
+          return;
+        }
+        ScopeTrackerScopeState scope(timer->object_, timer->dispatcher_);
```
Comment (mattklein123) [inline source/common/event/timer_impl.cc]: This change looks good to me, but do we have any tests that explicitly cover this? I might be missing it in the large test change.

Comment (alyssawilk) [reply source/common/event/timer_impl.cc]: No, I totally lost it during the rebase, thanks for catching that.
Thankfully where my local state was destroyed, github saved it for me :-)
https://github.com/envoyproxy/envoy/commit/ce642b2e501e92b53d5d49acb77c8696d514d721

**Pull request merged**